file_formats/cmt_format.ml

```diff
@@ -86,17 +85,113 @@ let need_to_clear_env =
 
 let keep_only_summary = Env.keep_only_summary
 
-open Tast_mapper
 
+let uid_to_loc : uid_fragment Types.Uid.Tbl.t ref =
+  Local_store.s_table Types.Uid.Tbl.create 16
 
+let register_uid uid fragment =
+  Types.Uid.Tbl.add !uid_to_loc uid fragment
 
 let rec tast_map =
+  let open Tast_mapper in
   let env_mapper =
-    if need_to_clear_env then (fun _sub env -> keep_only_summary env)
-    else Tast_mapper.default.env
+    { default with env =
+        if need_to_clear_env then (fun _sub env -> keep_only_summary env)
+        else Tast_mapper.default.env }
   in
-  fun ~env:_ () ->
-  {Tast_mapper.default with env = env_mapper }
+  fun ~env () -> { env_mapper with
+
+  signature = (fun _sub ({sig_final_env; _} as x) ->
+    let sub = tast_map ~env:sig_final_env () in
+    let sig_final_env = sub.env sub x.sig_final_env in
+    let sig_items = List.map (sub.signature_item sub) x.sig_items in
+    {x with sig_items; sig_final_env});
+
+  structure = (fun _sub {str_items; str_type; str_final_env} ->
+    let sub = tast_map ~env:str_final_env () in
+    {
+      str_items = List.map (sub.structure_item sub) str_items;
+      str_final_env = sub.env sub str_final_env;
+      str_type;
+    });
```

Comment:
Why do you need the `signature` and `structure` operations here? From a distance it looks like they only `map` without doing anything else special, so it should not be necessary to write them explicitly. If they are subtly doing something different from the default case that is important, they would deserve a comment.

Reply:
They actually do something: they update the environment in the closure of the mapper for the nested `Env.find` call to succeed. Since we are only interested in struct/sig level definitions / declarations we only need the "final" envs.

Another option would have been to map on the `struct_item` / `sig_item` cases to get more local environments. Not sure if it would be cleaner ?

Reply:
Ah, so you need an iterator with a view of the "current" typing environment. Maybe this can be made available globally instead of making it a specific change locally, but I'm not familiar with the typedtree-iterators stuff so I don't know offhand which API to suggest.

Reply:
I guess we could provide a small wrapper over the default iterators ?
but maybe maybe it is a bit out-of-scope for this PR...

Reply:
I think it would be much better to not use environments in this code. It looks like you are just using it to get the `uid`s for definitions from the typedtree. Those should really already be stored in the typedtree as part of those definitions.

Part of the point of `uid`s is to let people examine the binding structure of the tree without needing to mess around with environments.